base/compiler/abstractinterpretation.jl

```diff
@@ -34,34 +34,50 @@ function matching_methods(@nospecialize(atype), cache::IdDict{Any, Tuple{Any, UI
     return ms
 end
 
+# arr should be small - guaranteed below by the union splitting limit
+function push_unique!(arr::Vector{T}, x::T) where T
+    for y in arr
+        y === x && return
+    end
+    push!(arr, x)
+end
+
 function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f), argtypes::Vector{Any}, @nospecialize(atype), sv::InferenceState,
                                   max_methods::Int = InferenceParams(interp).MAX_METHODS)
     if sv.currpc in sv.throw_blocks
         return Any
     end
-    mt = ccall(:jl_method_table_for, Any, (Any,), atype)
-    if mt === nothing
-        add_remark!(interp, sv, "Could not identify method table for call")
-        return Any
-    end
-    mt = mt::Core.MethodTable
     min_valid = UInt[typemin(UInt)]
     max_valid = UInt[typemax(UInt)]
     atype_params = unwrap_unionall(atype).parameters
     splitunions = 1 < countunionsplit(atype_params) <= InferenceParams(interp).MAX_UNION_SPLITTING
     if splitunions
         splitsigs = switchtupleunion(atype)
         applicable = Any[]
+        mts = Core.MethodTable[]
         for sig_n in splitsigs
+            mt = ccall(:jl_method_table_for, Any, (Any,), sig_n)
+            if mt === nothing
+                add_remark!(interp, sv, "Could not identify method table for call")
+                return Any
+            end
+            mt = mt::Core.MethodTable
             xapplicable = matching_methods(sig_n, sv.matching_methods_cache, max_methods,
                                            get_world_counter(interp), min_valid, max_valid)
             if xapplicable === false
                 add_remark!(interp, sv, "For one of the union split cases, too many methods matched")
                 return Any
             end
             append!(applicable, xapplicable)
+            push_unique!(mts, mt)
```

Comment:
```suggestion
            contains_is(mts, mt) || push!(mts, mt)
```
(this is the pattern we use elsewhere anyways)